src/gadget/is_zero.rs

```diff
@@ -0,0 +1,450 @@
+        arithmetic::FieldExt,
+        circuit::{layouter::SingleChipLayouter, Layouter},
+        dev::{MockProver, VerifyFailure::Constraint},
+        plonk::{Advice, Assignment, Circuit, Column, ConstraintSystem, Error},
+        poly::Rotation,
+    };
+    use pasta_curves::pallas::Base;
+    use std::marker::PhantomData;
+
+    macro_rules! try_test_circuit {
+        ($values:expr, $checks:expr, $result:expr) => {{
+            let k = usize::BITS - $values.len().leading_zeros();
+            let circuit = TestCircuit::<Base> {
+                values: Some($values),
+                checks: Some($checks),
+                _marker: PhantomData,
+            };
+            let prover = MockProver::<Base>::run(k, &circuit, vec![]).unwrap();
+            assert_eq!(prover.verify(), $result);
+        }};
+    }
+
+    macro_rules! error_constraint_at_row {
+        ($row:expr) => {
+            Constraint {
+                gate_index: 1,
+                gate_name: "check is_zero",
+                constraint_index: 0,
+                constraint_name: "",
+                row: $row,
+            }
+        };
+    }
+
+    #[test]
+    fn row_diff_is_zero() {
+        #[derive(Clone, Debug)]
+        struct TestCircuitConfig<F> {
+            value: Column<Advice>,
+            check: Column<Advice>,
+            is_zero: IsZeroConfig<F>,
+        }
+
+        struct TestCircuit<F: FieldExt> {
+            values: Option<Vec<u64>>,
+            // checks[i] = is_zero(values[i + 1] - values[i])
+            checks: Option<Vec<bool>>,
+            _marker: PhantomData<F>,
+        }
+
+        impl<F: FieldExt> Circuit<F> for TestCircuit<F> {
+            type Config = TestCircuitConfig<F>;
+
+            fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+                let q_enable = meta.selector();
+                let value = meta.advice_column();
+                let value_diff_inv = meta.advice_column();
+                let check = meta.advice_column();
+
+                let is_zero = IsZeroChip::configure(
+                    meta,
+                    q_enable,
+                    |meta| {
+                        let value_prev = meta.query_advice(value, Rotation::prev());
+                        let value_cur = meta.query_advice(value, Rotation::cur());
+                        value_cur - value_prev
+                    },
+                    value_diff_inv,
+                );
+
+                let config = Self::Config {
+                    value,
+                    check,
+                    is_zero,
+                };
+
+                meta.create_gate("check is_zero", |meta| {
+                    let q_enable = meta.query_selector(q_enable);
+
+                    // This verifies is_zero is calculated correctly
+                    let check = meta.query_advice(config.check, Rotation::cur());
+
+                    vec![q_enable * (config.is_zero.is_zero_expression.clone() - check)]
```

Comment:
```suggestion
                    vec![q_enable * (config.is_zero.is_zero_expression - check)]
```

Reply:
The `config` here will be returned, so if in closure we still need `.clone()` (same as the above suggestion)